File: 22.linq-workshop-kata/ts/src/03-real-world/watch/component.ts
import { ComponentType, toComponentType } from "./component-type";

export interface RawComponent {
  name: string;
  id: string;
  type: string;
  components: RawComponent[];
}

export interface Component {
  readonly name: string;
  readonly id: string;
  readonly type: ComponentType;
  readonly components: ReadonlyArray<Component>;
}

export function parseComponent(raw: RawComponent): Component {
  return {
    name: raw.name,
    id: raw.id,
    type: toComponentType(raw.type),
    components: raw.components.map(parseComponent),
  };
}

File: 22.linq-workshop-kata/c#/03-real-world/Watch/Component.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace LinqWorkshop.Tests.RealWorld.Watch;

public class Component
{
    public string Name { get; set; } = "";
    public string Id { get; set; } = "";

    [JsonConverter(typeof(ComponentTypeConverter))]
    public ComponentType Type { get; set; }

    public List<Component> Components { get; set; } = [];
}
